src/cljc/schema/core.cljc

```diff
@@ -1398,10 +1398,28 @@
 (defmacro letfn
   "s/letfn : s/fn :: clojure.core/letfn : clojure.core/fn"
   [fnspecs & body]
-  (list `let
-        (vec (interleave (map first fnspecs)
-                         (map #(cons `fn %) fnspecs)))
-        `(do ~@body))))
+  (let [{:keys [outer-bindings
+                fnspecs
+                declare-class-schemas]}
+        (reduce (fn [acc fnspec]
+                  (let [[name more-fn-args] (macros/extract-arrow-schematized-element &env fnspec)
+                        {:keys [outer-bindings schema-form fn-body]} (macros/process-fn- &env name more-fn-args)]
+                    (-> acc
+                        (update :outer-bindings into outer-bindings)
+                        (update :fnspecs conj (cons name fn-body))
+                        (update :declare-class-schemas conj `(utils/declare-class-schema!
+                                                               (utils/fn-schema-bearer ~name)
+                                                               ~schema-form)))))
+                {:outer-bindings []
+                 :fnspecs []
+                 :declare-class-schemas []}
+                fnspecs)]
+    `(let ~outer-bindings
+       (clojure.core/letfn
+         ~fnspecs
+         (do
+           ~@declare-class-schemas
```

Comment:
My memory is a little rusty, but I think this is mutating some global variable, which seems weird.    I guess you get what you ask for if you put a `defn` or `defrecord` inside a `letfn` though, so this is probably the least surprising option?

Reply:
Yes, it's mutating a weak hash map. I went back and forth about this, since the `fn-schema` cannot be seen in the bindings if you use metadata.

eg.,
```
(s/fn-schema (letfn [(foo [] foo)] (foo)))
```

However one of my assumptions I just realized was wrong: a fn can't see it's own metadata.

```
user=> ((with-meta (fn this [] (meta this)) {:a 1}))
nil
```

So `s/fn` has the same problem. I think the metadata impl makes more sense to me now.